lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+      // If we're in the load mode that prefers .swiftinterfaces, specifically
+      // skip the module adjacent to the interface, but use the caches if
+      // they're present.
+      shouldLoadAdjacentModule = false;
+      break;
+    case ModuleLoadingMode::PreferSerialized:
+      // The rest of the function should be covered by this.
+      break;
+    case ModuleLoadingMode::OnlySerialized:
+      llvm_unreachable("parseable module loader should not have been created");
+    }
+
+    // First, check the cached module path. Whatever's in this cache represents
+    // the most up-to-date knowledge we have about the module.
+    if (fs.exists(cachedOutputPath)) {
+      auto bufOrErr = fs.getBufferForFile(cachedOutputPath);
+
+      // If we're unable to read this corrupted file, delete it from the
+      // cache and try to load again.
+      if (!bufOrErr) {
+        // FIXME: If this deletion fails, that's a fatal error. Ask the user
+        //        to delete the file?
+        llvm::sys::fs::remove(cachedOutputPath);
+        return discoverUpToDateModuleForInterface(modulePath,
+                                                  cachedOutputPath,
+                                                  deps);
+      }
+      auto &buf = *bufOrErr.get();
+
+      // Check to see if the module is a serialized AST. If it's not, then we're
+      // probably dealing with a Forwarding Module, which is a YAML file.
+      bool isForwardingModule =
+        !serialization::isSerializedAST(buf.getBuffer());
+
+      // If it's a forwarding module, load the YAML file from disk and check
+      // if it's up-to-date.
+      if (isForwardingModule) {
+        auto modOrErr = ForwardingModule::load(buf);
+        if (modOrErr && forwardingModuleIsUpToDate(*modOrErr, deps))
+          return DiscoveredModule::forwarded(modOrErr->underlyingModulePath);
+      // Otherwise, check if the AST buffer itself is up to date.
+      } else if (serializedASTBufferIsUpToDate(buf, deps)) {
+        return DiscoveredModule::normal(cachedOutputPath);
+      }
+    }
+
+    // If we have a prebuilt cache path, check that too if the interface comes
+    // from the SDK.
+    if (!prebuiltCacheDir.empty()) {
+      llvm::SmallString<256> scratch;
+      auto path = computePrebuiltModulePath(scratch);
+      if (path && swiftModuleIsUpToDate(*path, deps))
+        return DiscoveredModule::prebuilt(*path);
+    }
+
+    // Finally, if there's a module adjacent to the .swiftinterface that we can
+    // _likely_ load (it validates OK and is up to date), bail early with
+    // errc::not_supported, so the next (serialized) loader in the chain will
+    // load it. Alternately, if there's a .swiftmodule present but we can't even
+    // read it (for whatever reason), we should let the other module loader
+    // diagnose it.
+    if (shouldLoadAdjacentModule && fs.exists(modulePath)) {
+      if (auto modBuf = getSerializedASTBuffer(fs, modulePath)) {
```

Comment:
Same thing here; we should get the error from attempting to open the file, not check if it exists separately.